File: ruby/sdl_play.rb
def get_device_id
    print "Enter the ID of the joystick device: "
    gets.chomp
  end

  require 'json'

def read_input_file(input_file)
  File.read(input_file)
end

def playback_input(device_id, events)
    events.each do |event|
      `sdl2-jstest --event #{device_id} --#{event['type']} #{event['value']}`
      sleep(event['duration'])
    end
  end

  def main
    input_file = ARGV[0]
    device_id = get_device_id
    events = JSON.parse(read_input_file(input_file))
    playback_input(device_id, events)
  end

File: perl/sdl_play.pl
#!/usr/bin/perl

use strict;
use warnings;
use SDL;
use JSON;

sub play_input {
    my ($device_id) = @_;
    my $input_file = "input_record.json";
    SDL::init(SDL_INIT_JOYSTICK);
    my $joystick = SDL::Joystick->new($device_id);
    my $events = decode_json(join("", <>));
    foreach my $event (@{$events}) {
        my $timestamp = $event->{"timestamp"};
        my $type = $event->{"type"};
        my $code = $event->{"code"};
        my $value = $event->{"value"};
        my $duration = $event->{"duration"};
        SDL::delay($duration * 1000);
        if ($type == SDL_JOYAXISMOTION) {
            $joystick->set_axis($code, $value);
        } elsif ($type == SDL_JOYBUTTONDOWN) {
            $joystick->button_down($code);
        } elsif ($type == SDL_JOYBUTTONUP) {
            $joystick->button_up($code);
        }
    }
    SDL::Joystick::close($joystick);
}

my $device_id = 0; # Change this to the desired device ID
play_input($device_id);